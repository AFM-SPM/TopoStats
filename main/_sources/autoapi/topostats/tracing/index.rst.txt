topostats.tracing
=================

.. py:module:: topostats.tracing

.. autoapi-nested-parse::

   
   Initialise tracing sub-module.
















   ..
       !! processed by numpydoc !!


Submodules
----------

.. toctree::
   :maxdepth: 1

   /autoapi/topostats/tracing/disordered_tracing/index
   /autoapi/topostats/tracing/dnacurvature/index
   /autoapi/topostats/tracing/nodestats/index
   /autoapi/topostats/tracing/ordered_tracing/index
   /autoapi/topostats/tracing/pruning/index
   /autoapi/topostats/tracing/skeletonize/index
   /autoapi/topostats/tracing/splining/index
   /autoapi/topostats/tracing/tracingfuncs/index


