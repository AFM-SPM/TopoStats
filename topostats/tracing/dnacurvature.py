"""Module for calculating curvature statistics."""
import logging

# from pathlib import Path
import numpy as np
from topostats.logs.logs import LOGGER_NAME

LOGGER = logging.getLogger(LOGGER_NAME)

# Disable white space before colon
# black and flake8 conflict https://black.readthedocs.io/en/stable/faq.html#why-are-flake8-s-e203-and-w503-violated
# noqa: E203


class Curvature:
    def __init__(
        self,
        molecule_coordinates: np.ndarray,
        circular: bool,
    ):
        """Initialise the class.

        Parameters
        ----------
        molecule_coordinates: np.ndarray
            Coordinates of the simplified splined trace of a molecule. These are returned by dnaTracing.
        circular: bool
        """
        self.molecule_coordinates = molecule_coordinates
        self.circular = circular
        self.n_points = len(molecule_coordinates)
        self.first_derivative = None
        self.second_derivative = None
        self.local_curvature = None

    def calculate_derivatives(self, edge_order: int = 1) -> None:
        """Find the curvature for an individual molecule.

        Parameters
        ----------
        molecule_coordinates: np.ndarray
            Coordinates of the simplified splined trace of a molecule. These are returned by dnaTracing.
        circular: bool
            Whether the molecule has been determined as being circular or not.
        edge_order: int
            Gradient is passed to numpy.gradient and Gradient is calculated using N-th order accurate differences at
            boundaries. Also used to expand the array by the necessary number of co-ordinates at either end to form a
            loop for the calculations.
        """
        # If circular we need gradients correctly calculated at the start and end and so the array has the number of
        # points used in np.gradient(edge_order) from the end attached to the start and the same from the start attached
        # to the end.
        edge_order_boundary = edge_order + 1
        if self.circular:
            coordinates = np.vstack(
                (
                    self.molecule_coordinates[-edge_order_boundary:],
                    self.molecule_coordinates,
                    self.molecule_coordinates[:edge_order_boundary],
                )
            )
        else:
            coordinates = self.molecule_coordinates
        self.first_derivative = np.gradient(coordinates, edge_order, axis=0)
        self.second_derivative = np.gradient(self.first_derivative, edge_order, axis=0)
        # Now trim the arrays back to the appropriate size
        if self.circular:
            self.first_derivative = self.first_derivative[edge_order_boundary:-edge_order_boundary]
            self.second_derivative = self.second_derivative[edge_order_boundary:-edge_order_boundary]

    # def _extract_coordinates(molecule) -> np.ndarray:
    #     """Extract the coordinates for the points"""
    #     curve = []
    #     contour = 0
    #     # for i in len(molecule):
    #     pass

    def _calculate_local_curvature(self) -> float:
        """Calculate the local curvature between points."""
        self.local_curvature = (
            (self.second_derivative[:, 0] * self.first_derivative[:, 1])
            - (self.first_derivative[:, 0] * self.second_derivative[:, 1])
        ) / ((self.first_derivative[:, 0] ** 2) + (self.first_derivative[:, 1] ** 2)) ** 1.5
