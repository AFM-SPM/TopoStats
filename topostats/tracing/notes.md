odd number of branches
emanating_branch_starts_by_node: {0: [array([6, 1]), array([7, 3]), array([8, 1])], 1: [array([ 6, 11]), array([7, 9]), array([ 8, 11])]}
whole skel graph: <class 'networkx.classes.graph.Graph'>
Graph with 41 nodes and 42 edges
comparing 0 to 1
shortest path length between [6 1] and [ 6 11] is 10
shortest path length between [6 1] and [7 9] is 8
shortest path length between [6 1] and [ 8 11] is 10
shortest path length between [7 3] and [ 6 11] is 8
shortest path length between [7 3] and [7 9] is 6
shortest path length between [7 3] and [ 8 11] is 8
shortest path length between [8 1] and [ 6 11] is 10
shortest path length between [8 1] and [7 9] is 8
shortest path length between [8 1] and [ 8 11] is 10
comparing 1 to 0
shortest path length between [ 6 11] and [6 1] is 10
shortest path length between [ 6 11] and [7 3] is 8
shortest path length between [ 6 11] and [8 1] is 10
shortest path length between [7 9] and [6 1] is 8
shortest path length between [7 9] and [7 3] is 6
shortest path length between [7 9] and [8 1] is 8
shortest path length between [ 8 11] and [6 1] is 10
shortest path length between [ 8 11] and [7 3] is 8
shortest path length between [ 8 11] and [8 1] is 10
shortest_node_dists: [[0. 6.]
 [6. 0.]]
shortest_dists_branch_idxs: [[[0 0]
  [1 1]]

[[1 1]
  [0 0]]]
shortest_dist_coords: [[[[0 0]
   [0 0]]

[[7 3]
   [7 9]]]

[[[7 9]
   [7 3]]

[[0 0]
   [0 0]]]]

[[0 0 0 0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0 0 0 0]
 [0 0 0 1 1 1 1 1 1 1 0 0 0]
 [0 0 1 0 0 0 0 0 0 0 1 0 0]
 [0 1 0 0 0 0 0 0 0 0 0 1 0]
 [0 1 0 0 0 0 0 0 0 0 0 1 0]
 [0 1 0 0 0 0 0 0 0 0 0 1 0]
 [0 0 3 3 3 3 3 3 3 3 3 0 0]
 [0 1 0 0 0 0 0 0 0 0 0 1 0]
 [0 1 0 0 0 0 0 0 0 0 0 1 0]
 [0 1 0 0 0 0 0 0 0 0 0 1 0]
 [0 1 0 0 0 0 0 0 0 0 0 1 0]
 [0 0 1 1 1 1 1 1 1 1 1 0 0]
 [0 0 0 0 0 0 0 0 0 0 0 0 0]]
