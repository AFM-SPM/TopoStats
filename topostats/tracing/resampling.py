"""functions for trace smoothing"""

import numpy.typing as npt
import numpy as np
from topostats.io import LoadScans
import matplotlib.pyplot as plt
from topostats.unet_masking import make_bounding_box_square, pad_bounding_box


def interpolate_between_two_points_distance(
    point1: npt.NDArray[np.float32], point2: npt.NDArray[np.float32], distance: np.float32
) -> npt.NDArray[np.float32]:
    """Interpolate between two points to create a new point at a set distance between the two.

    Parameters
    ----------
    point1 : npt.NDArray[np.float32]
        The first point.
    point2 : npt.NDArray[np.float32]
        The second point.
    distance : np.float32
        The distance to interpolate between the two points.

    Returns
    -------
    npt.NDArray[np.float32]
        The new point at the specified distance between the two points.
    """
    distance_between_points = np.linalg.norm(point2 - point1)
    assert (
        distance_between_points > distance
    ), f"distance between points is less than the desired interval: {distance_between_points} < {distance}"
    proportion = distance / distance_between_points
    new_point = point1 + proportion * (point2 - point1)
    return new_point


def resample_points_regular_interval(points: npt.NDArray, interval: float, circular: bool) -> npt.NDArray:
    """Resample a set of points to be at regular spatial intervals.
    
    Note: This is NOT intended to be pure interpolation, as interpolated points would not produce uniformly spaced points."""

    if circular:
        points = np.concatenate((points, points[0:1]), axis=0)

    resampled_points = []
    resampled_points.append(points[0])
    current_point_index = 1
    while True:
        current_point = resampled_points[-1]
        next_original_point = points[current_point_index]
        distance_to_next_splined_point = np.linalg.norm(next_original_point - current_point)
        # if the distance to the next splined point is less than the interval, then skip to the next point
        if distance_to_next_splined_point < interval:
            current_point_index += 1
            if current_point_index >= len(points):
                break
            continue
        new_interpolated_point = interpolate_between_two_points(current_point, next_original_point, interval)
        resampled_points.append(new_interpolated_point)

    # if the first and last points are less than 0.5 * the interval apart, then remove the last point
    if np.linalg.norm(resampled_points[0] - resampled_points[-1]) < 0.5 * interval:
        resampled_points = resampled_points[:-1]

    resampled_points_numpy = np.array(resampled_points)

    return resampled_points_numpy
