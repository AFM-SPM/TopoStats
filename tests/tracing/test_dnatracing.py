"""Tests of dnatracing methods."""
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import pytest

from topostats.tracing.dnatracing import dnaTrace, traceStats

# pylint: disable=protected-access


def test_get_grain_array(test_dnatracing: pd.DataFrame) -> None:
    """Test extraction of grain array."""
    target = {
        1: np.array(
            [
                [0, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0],
                [0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0],
                [0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0],
                [0, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0],
                [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
                [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
                [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
                [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
                [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
            ]
        ),
        2: np.array(
            [
                [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1],
                [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1],
                [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1],
                [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1],
                [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1],
                [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1],
                [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1],
                [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1],
                [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1],
            ]
        ),
        3: np.array(
            [
                [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
                [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
                [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
                [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
                [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
                [0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0],
                [0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0],
                [0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0],
                [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
            ]
        ),
        4: np.array(
            [
                [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
                [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
                [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
                [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
                [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
                [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
                [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
                [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
                [0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0],
            ]
        ),
    }
    for x in range(1, 4):
        np.testing.assert_array_equal(test_dnatracing._get_grain_array(x), target[x])


@pytest.mark.parametrize(
    "molecule, array_sum",
    [
        (1, 2872),
        (2, 1726),
        (3, 1770),
        (4, 1624),
        (5, 1706),
        (6, 1695),
        (7, 1729),
        (8, 1855),
        (9, 2085),
        (10, 1900),
        (11, 2054),
        (12, 2003),
        (13, 1919),
        (14, 1887),
        (15, 1766),
        (16, 1898),
        (17, 1965),
        (18, 2000),
        (19, 1818),
        (20, 1967),
        (21, 1895),
        (22, 1826),
    ],
)
def test_minicircle_dnatracing_skeleton_size(regtest, minicircle_dnatracing: traceStats, molecule, array_sum) -> None:
    """Test the skeletons returned by dnatracing."""
    assert minicircle_dnatracing._get_grain_array(molecule).sum() == array_sum
    mol_array = minicircle_dnatracing._get_grain_array(1)


def test_minicircle_dnatracing_skeletons(regtest, minicircle_dnatracing: traceStats) -> None:
    """Test the skeletons returned by dnatracing."""
    mol_array = minicircle_dnatracing._get_grain_array(1)
    assert isinstance(mol_array, np.ndarray)
    print(np.array2string(mol_array), file=regtest)


# @pytest.mark.mpl_image_compare(baseline_dir="../resources/img/")
# def test_minicircle_dnatracing_skeletons_plot(regtest, minicircle_dnatracing: traceStats) -> None:
#     """Test the skeletons returned by dnatracing."""
#     mol_array = minicircle_dnatracing._get_grain_array(1)
#     fig, ax = plt.subplots(1, 1, figsize=(8, 8))
#     ax.imshow(mol_array)
#     return fig


def test_tracestats(regtest, minicircle_tracestats: traceStats) -> None:
    """Regression tests for DNA trace statistics."""
    print(minicircle_tracestats.to_string(), file=regtest)
