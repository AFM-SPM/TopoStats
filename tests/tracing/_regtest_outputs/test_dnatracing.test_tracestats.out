                 contour_lengths  circular  end_to_end_distance image
molecule_number                                                      
0                   2.256071e-07     False         3.501502e-08   tmp
1                   4.582463e-08      True         0.000000e+00   tmp
2                   1.277585e-07     False         2.209248e-08   tmp
3                   9.458398e-08      True         0.000000e+00   tmp
4                   9.722231e-08     False         2.590102e-08   tmp
5                   9.579671e-08      True         0.000000e+00   tmp
6                   9.105686e-08     False         2.255165e-08   tmp
7                   9.889547e-08      True         0.000000e+00   tmp
8                   1.076075e-07     False         2.304406e-08   tmp
9                   5.692076e-08      True         0.000000e+00   tmp
10                  7.649301e-08      True         0.000000e+00   tmp
11                  1.040888e-07     False         2.739824e-08   tmp
12                  8.293111e-08      True         0.000000e+00   tmp
13                  1.016983e-07      True         0.000000e+00   tmp
14                  1.001398e-07      True         0.000000e+00   tmp
15                  6.457170e-08      True         0.000000e+00   tmp
16                  9.529778e-08     False         2.745163e-08   tmp
17                  1.070051e-07      True         0.000000e+00   tmp
18                  7.138138e-08      True         0.000000e+00   tmp
19                  1.016008e-07      True         0.000000e+00   tmp
20                  1.079293e-07     False         1.635441e-08   tmp
21                  9.340060e-08     False         2.006041e-08   tmp
